rosnode - object_detection_2d_yolov3.py - error when echoing objects topic

The output image is working but I cannot echo the object topic, I am using opendr installed on my computer on the develop branch

First I get this warning:
```
/home/user/opendr/venv/lib/python3.8/site-packages/numpy/core/fromnumeric.py:3432: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
/home/user/opendr/venv/lib/python3.8/site-packages/numpy/core/_methods.py:190: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
/home/user/opendr/src/opendr/perception/object_detection_2d/datasets/transforms.py:34: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  bbox_np = np.asarray([bbox.left, bbox.top, bbox.left + bbox.width, bbox.top + bbox.height, bbox.confidence, bbox.name])
```
And then I get this error when trying to echo the /open/objects:
`rostopic echo /open/objects`

```
[ERROR] [1662986163.583257]: bad callback: <bound method ObjectDetectionYOLONode.callback of <__main__.ObjectDetectionYOLONode object at 0x7f274299fc10>>
Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/vision_msgs/msg/_Detection2DArray.py", line 247, in serialize
    buff.write(_get_struct_qd().pack(_x.id, _x.score))
TypeError: only integer scalar arrays can be converted to a scalar index

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 882, in publish
    self.impl.publish(data)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 1066, in publish
    serialize_message(b, self.seq, message)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/msg.py", line 152, in serialize_message
    msg.serialize(b)
  File "/opt/ros/noetic/lib/python3/dist-packages/vision_msgs/msg/_Detection2DArray.py", line 294, in serialize
    except TypeError as te: self._check_types(ValueError("%s: '%s' when writing '%s'" % (type(te), str(te), str(locals().get('_x', self)))))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 393, in _check_types
    check_type(n, t, getattr(self, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 314, in check_type
    check_type(field_name + '[]', base_type, v)
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 324, in check_type
    check_type('%s.%s' % (field_name, n), t, getattr(field_val, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 314, in check_type
    check_type(field_name + '[]', base_type, v)
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 324, in check_type
    check_type('%s.%s' % (field_name, n), t, getattr(field_val, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 261, in check_type
    raise SerializationError('field %s must be an integer type' % field_name)
genpy.message.SerializationError: field detections[].results[].id must be an integer type

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "/home/user/opendr/projects/opendr_ws/src/perception/scripts/object_detection_2d_yolov3.py", line 95, in callback
    self.object_publisher.publish(ros_boxes)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 886, in publish
    raise ROSSerializationException(str(e))
rospy.exceptions.ROSSerializationException: field detections[].results[].id must be an integer type

[ERROR] [1662986163.943623]: bad callback: <bound method ObjectDetectionYOLONode.callback of <__main__.ObjectDetectionYOLONode object at 0x7f274299fc10>>
Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/vision_msgs/msg/_Detection2DArray.py", line 247, in serialize
    buff.write(_get_struct_qd().pack(_x.id, _x.score))
TypeError: only integer scalar arrays can be converted to a scalar index

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 882, in publish
    self.impl.publish(data)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 1066, in publish
    serialize_message(b, self.seq, message)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/msg.py", line 152, in serialize_message
    msg.serialize(b)
  File "/opt/ros/noetic/lib/python3/dist-packages/vision_msgs/msg/_Detection2DArray.py", line 294, in serialize
    except TypeError as te: self._check_types(ValueError("%s: '%s' when writing '%s'" % (type(te), str(te), str(locals().get('_x', self)))))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 393, in _check_types
    check_type(n, t, getattr(self, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 314, in check_type
    check_type(field_name + '[]', base_type, v)
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 324, in check_type
    check_type('%s.%s' % (field_name, n), t, getattr(field_val, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 314, in check_type
    check_type(field_name + '[]', base_type, v)
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 324, in check_type
    check_type('%s.%s' % (field_name, n), t, getattr(field_val, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 261, in check_type
    raise SerializationError('field %s must be an integer type' % field_name)
genpy.message.SerializationError: field detections[].results[].id must be an integer type

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "/home/user/opendr/projects/opendr_ws/src/perception/scripts/object_detection_2d_yolov3.py", line 95, in callback
    self.object_publisher.publish(ros_boxes)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 886, in publish
    raise ROSSerializationException(str(e))
rospy.exceptions.ROSSerializationException: field detections[].results[].id must be an integer type

[ERROR] [1662986164.310993]: bad callback: <bound method ObjectDetectionYOLONode.callback of <__main__.ObjectDetectionYOLONode object at 0x7f274299fc10>>
Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/vision_msgs/msg/_Detection2DArray.py", line 247, in serialize
    buff.write(_get_struct_qd().pack(_x.id, _x.score))
TypeError: only integer scalar arrays can be converted to a scalar index

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 882, in publish
    self.impl.publish(data)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 1066, in publish
    serialize_message(b, self.seq, message)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/msg.py", line 152, in serialize_message
    msg.serialize(b)
  File "/opt/ros/noetic/lib/python3/dist-packages/vision_msgs/msg/_Detection2DArray.py", line 294, in serialize
    except TypeError as te: self._check_types(ValueError("%s: '%s' when writing '%s'" % (type(te), str(te), str(locals().get('_x', self)))))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 393, in _check_types
    check_type(n, t, getattr(self, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 314, in check_type
    check_type(field_name + '[]', base_type, v)
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 324, in check_type
    check_type('%s.%s' % (field_name, n), t, getattr(field_val, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 314, in check_type
    check_type(field_name + '[]', base_type, v)
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 324, in check_type
    check_type('%s.%s' % (field_name, n), t, getattr(field_val, n))
  File "/opt/ros/noetic/lib/python3/dist-packages/genpy/message.py", line 261, in check_type
    raise SerializationError('field %s must be an integer type' % field_name)
genpy.message.SerializationError: field detections[].results[].id must be an integer type

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "/home/user/opendr/projects/opendr_ws/src/perception/scripts/object_detection_2d_yolov3.py", line 95, in callback
    self.object_publisher.publish(ros_boxes)
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 886, in publish
    raise ROSSerializationException(str(e))
rospy.exceptions.ROSSerializationException: field detections[].results[].id must be an integer type
```

--- a/projects/opendr_ws/src/ros_bridge/src/opendr_bridge/bridge.py
+++ b/projects/opendr_ws/src/ros_bridge/src/opendr_bridge/bridge.py
@@ -294,7 +294,7 @@ class ROSBridge:
             detection.bbox.center.y = bounding_box.top + bounding_box.height / 2.0
             detection.bbox.size_x = bounding_box.width
             detection.bbox.size_y = bounding_box.height
-            detection.results[0].id = bounding_box.name
+            detection.results[0].id = int(bounding_box.name)
             detection.results[0].score = bounding_box.confidence
             detections.detections.append(detection)
         return detections

--- a/src/opendr/perception/object_detection_2d/yolov3/yolov3_learner.py
+++ b/src/opendr/perception/object_detection_2d/yolov3/yolov3_learner.py
@@ -572,21 +572,38 @@ class YOLOv3DetectorLearner(Learner):
                                     "yolo_voc.json")
             if verbose:
                 print("Downloading metadata...")
-            urlretrieve(file_url, os.path.join(path, "yolo_default.json"))
+            if not os.path.exists(os.path.join(path, "yolo_default.json")):
+                urlretrieve(file_url, os.path.join(path, "yolo_default.json"))
+                if verbose:
+                    print("Downloaded metadata json.")
+            else:
+                if verbose:
+                    print("Metadata json file already exists.")
 
             if verbose:
                 print("Downloading params...")
             file_url = os.path.join(url, "pretrained", "yolo_voc",
                                          "yolo_voc.params")
 
-            urlretrieve(file_url,
-                        os.path.join(path, "yolo_voc.params"))
+            if not os.path.exists(os.path.join(path, "yolo_voc.params")):
+                urlretrieve(file_url, os.path.join(path, "yolo_voc.params"))
+                if verbose:
+                    print("Downloaded params.")
+            else:
+                if verbose:
+                    print("Params file already exists.")
 
         elif mode == "images":
             file_url = os.path.join(url, "images", "cat.jpg")
             if verbose:
                 print("Downloading example image...")
-            urlretrieve(file_url, os.path.join(path, "cat.jpg"))
+            if not os.path.exists(os.path.join(path, "cat.jpg")):
+                urlretrieve(file_url, os.path.join(path, "cat.jpg"))
+                if verbose:
+                    print("Downloaded example image.")
+            else:
+                if verbose:
+                    print("Example image already exists.")
 
         elif mode == "test_data":
             os.makedirs(os.path.join(path, "test_data"), exist_ok=True)
@@ -596,17 +613,35 @@ class YOLOv3DetectorLearner(Learner):
             file_url = os.path.join(url, "test_data", "train.txt")
             if verbose:
                 print("Downloading filelist...")
-            urlretrieve(file_url, os.path.join(path, "test_data", "train.txt"))
+            if not os.path.exists(os.path.join(path, "test_data", "train.txt")):
+                urlretrieve(file_url, os.path.join(path, "test_data", "train.txt"))
+                if verbose:
+                    print("Downloaded filelist.")
+            else:
+                if verbose:
+                    print("Filelist already exists.")
             # download image
             file_url = os.path.join(url, "test_data", "Images", "000040.jpg")
             if verbose:
                 print("Downloading image...")
-            urlretrieve(file_url, os.path.join(path, "test_data", "Images", "000040.jpg"))
+            if not os.path.exists(os.path.join(path, "test_data", "Images", "000040.jpg")):
+                urlretrieve(file_url, os.path.join(path, "test_data", "Images", "000040.jpg"))
+                if verbose:
+                    print("Downloaded image.")
+            else:
+                if verbose:
+                    print("Image already exists.")
             # download annotations
             file_url = os.path.join(url, "test_data", "Annotations", "000040.jpg.txt")
             if verbose:
                 print("Downloading annotations...")
-            urlretrieve(file_url, os.path.join(path, "test_data", "Annotations", "000040.jpg.txt"))
+            if not os.path.exists(os.path.join(path, "test_data", "Annotations", "000040.jpg.txt")):
+                urlretrieve(file_url, os.path.join(path, "test_data", "Annotations", "000040.jpg.txt"))
+                if verbose:
+                    print("Downloaded annotations.")
+            else:
+                if verbose:
+                    print("Annotations already exist.")
 
     def optimize(self, target_device):
         """This method is not used in this implementation."""

--- a/src/opendr/perception/object_detection_2d/yolov3/yolov3_learner.py
+++ b/src/opendr/perception/object_detection_2d/yolov3/yolov3_learner.py
@@ -576,9 +576,8 @@ class YOLOv3DetectorLearner(Learner):
                 urlretrieve(file_url, os.path.join(path, "yolo_default.json"))
                 if verbose:
                     print("Downloaded metadata json.")
-            else:
-                if verbose:
-                    print("Metadata json file already exists.")
+            elif verbose:
+                print("Metadata json file already exists.")
 
             if verbose:
                 print("Downloading params...")
@@ -589,9 +588,8 @@ class YOLOv3DetectorLearner(Learner):
                 urlretrieve(file_url, os.path.join(path, "yolo_voc.params"))
                 if verbose:
                     print("Downloaded params.")
-            else:
-                if verbose:
-                    print("Params file already exists.")
+            elif verbose:
+                print("Params file already exists.")
 
         elif mode == "images":
             file_url = os.path.join(url, "images", "cat.jpg")
@@ -601,9 +599,8 @@ class YOLOv3DetectorLearner(Learner):
                 urlretrieve(file_url, os.path.join(path, "cat.jpg"))
                 if verbose:
                     print("Downloaded example image.")
-            else:
-                if verbose:
-                    print("Example image already exists.")
+            elif verbose:
+                print("Example image already exists.")
 
         elif mode == "test_data":
             os.makedirs(os.path.join(path, "test_data"), exist_ok=True)
@@ -617,9 +614,8 @@ class YOLOv3DetectorLearner(Learner):
                 urlretrieve(file_url, os.path.join(path, "test_data", "train.txt"))
                 if verbose:
                     print("Downloaded filelist.")
-            else:
-                if verbose:
-                    print("Filelist already exists.")
+            elif verbose:
+                print("Filelist already exists.")
             # download image
             file_url = os.path.join(url, "test_data", "Images", "000040.jpg")
             if verbose:
@@ -628,9 +624,8 @@ class YOLOv3DetectorLearner(Learner):
                 urlretrieve(file_url, os.path.join(path, "test_data", "Images", "000040.jpg"))
                 if verbose:
                     print("Downloaded image.")
-            else:
-                if verbose:
-                    print("Image already exists.")
+            elif verbose:
+                print("Image already exists.")
             # download annotations
             file_url = os.path.join(url, "test_data", "Annotations", "000040.jpg.txt")
             if verbose:
@@ -639,9 +634,8 @@ class YOLOv3DetectorLearner(Learner):
                 urlretrieve(file_url, os.path.join(path, "test_data", "Annotations", "000040.jpg.txt"))
                 if verbose:
                     print("Downloaded annotations.")
-            else:
-                if verbose:
-                    print("Annotations already exist.")
+            elif verbose:
+                print("Annotations already exist.")
 
     def optimize(self, target_device):
         """This method is not used in this implementation."""

